packager/packager.cc

```diff
@@ -1560,71 +1560,91 @@ unique_ptr<PackageInfoImpl> createAndPopulatePackageInfo(core::GlobalState &gs,
 // https://www.cs.cmu.edu/~15451-f18/lectures/lec19-DFS-strong-components.pdf provides a good overview of the
 // algorithm.
 const int UNVISITED = 0;
-struct ComputeSCCsMetadata {
-    int nextIndex = 1;
-    int nextSCCId = 0;
+struct NodeInfo {
     // A given package's index in the DFS traversal; ie. when it was first visited. The default value of 0 means the
     // package hasn't been visited yet.
-    UnorderedMap<core::packages::MangledName, int> index;
+    int index = UNVISITED;
     // The lowest index reachable from a given package (in the same SCC) by following any number of tree edges
     // and at most one back/cross edge
-    UnorderedMap<core::packages::MangledName, int> lowLink;
+    int lowLink = 0;
     // Fast way to check if a package is on the stack
-    UnorderedMap<core::packages::MangledName, bool> onStack;
+    bool onStack = false;
+};
+
+struct ComputeSCCsMetadata {
+    int nextIndex = 1;
+    int nextSCCId = 0;
+    UnorderedMap<core::packages::MangledName, NodeInfo> nodeMap;
     // As we visit packages, we push them onto the stack. Once we find the "root" of an SCC, we can use the stack to
     // determine all packages in the SCC.
     std::vector<core::packages::MangledName> stack;
 };
 
 // DFS traversal for Tarjan's algorithm starting from pkgName, along with keeping track of some metadata needed for
 // detecting SCCs.
-void strongConnect(core::GlobalState &gs, ComputeSCCsMetadata &metadata, core::packages::MangledName pkgName) {
+void strongConnect(core::GlobalState &gs, ComputeSCCsMetadata &metadata, core::packages::MangledName pkgName,
+                   NodeInfo &infoAtEntry) {
     auto *pkgInfoPtr = gs.packageDB().getPackageInfoNonConst(pkgName);
     if (!pkgInfoPtr) {
         // This is to handle the case where the user imports a package that doesn't exist.
         return;
     }
     auto &pkgInfo = PackageInfoImpl::from(*pkgInfoPtr);
-    metadata.index[pkgName] = metadata.nextIndex;
-    metadata.lowLink[pkgName] = metadata.nextIndex;
-    metadata.nextIndex++;
-    metadata.stack.push_back(pkgName);
-    metadata.onStack[pkgName] = true;
+    {
+        infoAtEntry.index = metadata.nextIndex;
+        infoAtEntry.lowLink = metadata.nextIndex;
+        metadata.nextIndex++;
+        metadata.stack.push_back(pkgName);
+        infoAtEntry.onStack = true;
+    }
     for (auto &i : pkgInfo.importedPackageNames) {
         if (i.type == ImportType::Test) {
             continue;
         }
-        if (metadata.index[i.name.mangledName] == UNVISITED) {
+        // We need to be careful with this; it's not valid after a call to `strongConnect`,
+        // because our reference might disappear from underneath us during that call.
```

Comment:
Could we eagerly populate the map to avoid the potential invalidation?

Reply:
I wondered if the `reserve` call would effectively do this, but I think it doesn't work: we can have (erroneous) imports that don't exist in the package database but do get entries in the hash table, and we'll only discover those via traversal and (implicit) insertions.  So even if we pre-reserve the space and eagerly populate the map with entries for all the known packages, we can still wind up invalidating the map's storage.

cc @neilparikh who knows this code better than I do to confirm that this intuition is correct.

Reply:
Thanks for confirming.

I don't think we know _a priori_ how many invalid imports we would encounter, and I wouldn't like to build assumptions about "oh, nobody would have more than 10 invalid imports" or whatever into Sorbet.

Reply:
Ya, that's reasonable.

I think what might be worth doing is creating an iterator that just returns the valid imports in a package and then this code (as well as other other that validates them) wouldn't have to worry about the non-existent packages.

Doesn't have be done in this PR though, I'll look into that refactor separately later.